R/retry.R

```diff
@@ -0,0 +1,58 @@
+#' Retry a request until it succeeds.
+#'
+#' Safely retry a request until it succeeds (returns an HTTP status code
+#' below 400). It is designed to be kind to the server: after each failure
+#' randomly waits up to twice as long. (Technically it uses exponential
+#' backoff with jitter, using the approach outlined in
+#' \url{https://www.awsarchitectureblog.com/2015/03/backoff.html.})
+#'
+#' @inheritParams VERB
+#' @inheritParams GET
+#' @inheritParams POST
+#' @param times Maximum number of requests to attempt.
+#' @param pause_base,pause_cap This method uses exponential back-off of with
+#'   full jitter - this means that each request will randomly wait between 0
+#'   and \code{pause_base * 2 ^ attempt} seconds, up to a maximum of
+#'   \code{pause_cap} seconds.
+#' @param quiet If \code{FALSE}, will print a message displaying how long
+#'   until the next request.
+#' @return The last response. Note that if the request doesn't succeed after
+#'   \code{times} times this will be a failed request, i.e. you still need
+#'   to use \code{\link{stop_for_status}()}.
+#' @export
+#' @examples
+#' # Succeeds straight away
+#' RETRY("GET", "http://httpbin.org/status/200")
+#' # Never succeeds
+#' RETRY("GET", "http://httpbin.org/status/404")
+RETRY <- function(verb, url = NULL, config = list(), ...,
+                  body = NULL, encode = c("multipart", "form", "json", "raw"),
+                  times = 3, pause_base = 1, pause_cap = 60,
+                  handle = NULL, quiet = FALSE) {
+  stopifnot(is.numeric(times), length(times) == 1L)
+
+  hu <- handle_url(handle, url, ...)
+  req <- request_build(verb, hu$url, body_config(body, match.arg(encode)), config, ...)
+  resp <- request_perform(req, hu$handle$handle)
```

Comment:
But don't you need an additional check? `i <= times && (is.null(resp) || http_error(resp))` ?   (I think you actually want a do-while loop here, but R doesn't have that)